src/state_lexcounter.erl

```diff
@@ -0,0 +1,243 @@
+
+-behaviour(type).
+-behaviour(state_type).
+
+-define(TYPE, ?MODULE).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, delta_mutate/3, merge/2]).
+-export([query/1, equal/2, is_inflation/2, is_strict_inflation/2]).
+-export([join_decomposition/1]).
+
+-export_type([state_lexcounter/0, delta_state_lexcounter/0, state_lexcounter_op/0]).
+
+-opaque state_lexcounter() :: {?TYPE, payload()}.
+-opaque delta_state_lexcounter() :: {?TYPE, {delta, payload()}}.
+-type delta_or_state() :: state_lexcounter() | delta_state_lexcounter().
+-type payload() :: orddict:orddict().
+-type state_lexcounter_op() :: increment | decrement.
+
+%% @doc Create a new, empty `state_lexcounter()'
+-spec new() -> state_lexcounter().
+new() ->
+    {?TYPE, orddict:new()}.
+
+%% @doc Create a new, empty `state_lexcounter()'
+-spec new([term()]) -> state_lexcounter().
+new([]) ->
+    new().
+
+%% @doc Mutate a `state_lexcounter()'.
+-spec mutate(state_lexcounter_op(), type:id(), state_lexcounter()) ->
+    {ok, state_lexcounter()}.
+mutate(Op, Actor, {?TYPE, _LexCounter}=CRDT) ->
+    state_type:mutate(Op, Actor, CRDT).
+
+%% @doc Delta-mutate a `state_lexcounter()'.
+%%      The first argument can be `increment' or `decrement'.
+%%      The second argument is the replica id.
+%%      The third argument is the `state_lexcounter()' to be inflated.
+-spec delta_mutate(state_lexcounter_op(), type:id(), state_lexcounter()) ->
+    {ok, delta_state_lexcounter()}.
+delta_mutate(increment, Actor, {?TYPE, LexCounter}) ->
+    {Left, Right} = case orddict:find(Actor, LexCounter) of
+        {ok, Value} ->
+            Value;
+        error ->
+            {0, 0}
+    end,
+    Delta = orddict:store(Actor, {Left, Right + 1}, orddict:new()),
+    {ok, {?TYPE, {delta, Delta}}};
+
+delta_mutate(decrement, Actor, {?TYPE, LexCounter}) ->
+    {Left, Right} = case orddict:find(Actor, LexCounter) of
+        {ok, Value} ->
+            Value;
+        error ->
+            {0, 0}
+    end,
+    Delta = orddict:store(Actor, {Left + 1, Right - 1}, orddict:new()),
+    {ok, {?TYPE, {delta, Delta}}}.
+
+%% @doc Returns the value of the `state_lexcounter()'.
+%%      A `state_lexcounter()' is a dictionary where the values are
+%%      pairs. The value of the `state_lexcounter()' is the sum of
+%%      the second components of these pairs.
+-spec query(state_lexcounter()) -> non_neg_integer().
+query({?TYPE, LexCounter}) ->
+    lists:sum([ Right || {_Actor, {_Left, Right}} <- LexCounter ]).
+
+%% @doc Merge two `state_lexcounter()'.
+%%      The keys of the resulting `state_lexcounter()' are the union of the
+%%      keys of both `state_lexcounter()' passed as input.
+%%      If a key is only present on one of the `state_lexcounter()',
+%%      its correspondent value is preserved.
+%%      If a key is present in both `state_lexcounter()', the new value
+%%      will be the lexicographic join of both values.
```

Comment:
In this case "the join of the lexicographic pairs" looks better since here there are not two different joins for pairs.